README.md
lagotto-rb
==========

[![Build Status](https://api.travis-ci.org/lagotto/lagotto-rb.png)](https://travis-ci.org/lagotto/lagotto-rb)
[![gem version](https://img.shields.io/gem/v/lagotto-rb.svg)](https://rubygems.org/gems/lagotto-rb)
[![codecov.io](http://codecov.io/github/lagotto/lagotto-rb/coverage.svg?branch=master)](http://codecov.io/github/lagotto/lagotto-rb?branch=master)

A Ruby client for the Lagotto application for article level metrics data

__This is alpha software, so expect changes__

Other Lagotto clients:

* R - [alm](https://github.com/ropensci/alm)
* Python - [pyalm](https://github.com/lagotto/pyalm)

## Changes

For changes see the [Changelog][changelog]

## API

Methods in relation to [Lagotto API][lapi] routes

* `/works` - `Lagotto.works()`
* `/works_sources` - `Lagotto.works_sources()`
* `/events` - `Lagotto.events()`
* `/publishers` - `Lagotto.publishers()`
* `/groups` - `Lagotto.groups()`
* `/references` - `Lagotto.references()`
* `/work_types` - `Lagotto.work_types()`
* `/docs` - `Lagotto.docs()`
* `/relation_types` - `Lagotto.relation_types()`
* `/sources` - `Lagotto.sources()`
* `/recommendations` - `Lagotto.recommendations()`
* `/status` - `Lagotto.status()`

## Install

### Release version

```
gem install lagotto-rb
```

### Development version

Install dependencies

```
git clone git@github.com:sckott/serrano.git
cd serrano
rake install
```

### In Ruby repl

Get altmetrics by DOI

```ruby
Lagotto.works(ids: 'http://doi.org/10.15468/DL.SQNY5P', instance: "crossref")
```

Search for altmetrics by source

```ruby
Lagotto.works(source: 'twitter', instance: "crossref")
```

Make a `/works` route request by source

```ruby
Lagotto.works_sources(source: 'twitter', per_page: 2)
```

Make a `/events` route request

```ruby
Lagotto.events(ids: '10.1371/journal.pone.0029797', instance: "crossref")
```

Make a `/publishers` route request

```ruby
Lagotto.publishers(id: 340)
```

Make a `/groups` route request

```ruby
Lagotto.groups()
```

Make a `/references` route request

```ruby
Lagotto.references(per_page: 5, instance: 'crossref')
```

### On the CLI

```
~$ lagotto
Commands:
  lagotto help [COMMAND]  # Describe available commands or one specific command
  lagotto version         # Get lagotto-rb version
  lagotto works [ids]     # Get works by ids
```

```
# A single id
~$ lagotto works http://doi.org/10.1371/journal.pone.0025110

DOI: 10.1371/journal.pone.0033693
type: journal-article
title: Methylphenidate Exposure Induces Dopamine Neuron Loss and Activation of Microglia in the Basal Ganglia of Mice

# JSON output
~$ lagotto works --limit=2 --json

{"meta":{"status":"ok","message-type":"work-list", ...

# JSON output, parse with jq
~$ lagotto works --limit=2 --json | jq .works[].DOI

"10.1371/journal.pgen.1005692"
"10.1371/journal.pgen.1005425"
```

## Meta

* Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.
* License: MIT

[lapi]: http://alm.plos.org/docs/api
[changelog]: https://github.com/lagotto/lagotto-rb/blob/master/CHANGELOG.md


CHANGELOG.md
## 0.2.0 (2016-01-08)

+ First version to Rubygems


lib/lagotto-rb.rb
require "multi_json"

require "lagotto-rb/version"
require "lagotto-rb/request"
require "lagotto-rb/req"
require "lagotto-rb/helpers"

# @!macro lagotto_params
#   @param ids [String] one or more work identifier
#   @param type [String] One of doi, pmid, pmcid, or mendeley_uuid
#   @param source [String] One source. To get many sources, make many calls.
#   @param publisher [String] Filter articles to a given publisher, using a crossref_id.
#   @param order [String] Results are sorted by descending event count when given the source name, e.g. `&order=wikipedia`. Otherwise (the default) results are sorted by date descending. When using `&source=x`, we can only sort by data or that source, not a different source.
#   @param per_page [String] Items per page
#   @param page [String] Page to retrieve
#   @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
#   @param key [String] API key
#   @param verbose [Boolean] Print request headers to stdout. Default: false

# @!macro lagotto_options
#   @param options [Hash] Hash of options for configuring the request, passed on to Faraday.new
#     :timeout      - [Fixnum] open/read timeout Integer in seconds
#     :open_timeout - [Fixnum] read timeout Integer in seconds
#     :proxy        - [Hash] hash of proxy options
#       :uri - [String] Proxy Server URI
#       :user - [String] Proxy server username
#       :password - [String] Proxy server password
#     :params_encoder - [Hash] not sure what this is
#     :bind           - [Hash] A hash with host and port values
#     :boundary       - [String] of the boundary value
#     :oauth          - [Hash] A hash with OAuth details

##
# Lagotto - The top level module for using methods
# to access a Lagotto instance API
#
# The following methods, matching the main Crossref API routes, are available:
# * `Lagotto.works` - Use the /works endpoint
# * `Lagotto.status` - Get instance status, requires authentication
# * `Lagotto.requests` - Search requests, requires authentication
# * `Lagotto.alerts` - Search alerts, requires authentication
# * `Lagotto.events` - Get events back
# * `Lagotto.sources` - Search /works routes by source
module Lagotto
  ##
  # Make a `/works` route request
  #
  # @!macro lagotto_params
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #      require 'lagotto-rb'
  #      Lagotto.works(ids: 'http://doi.org/10.15468/DL.SQNY5P', instance: "crossref")
  #      Lagotto.works(ids: ['http://doi.org/10.1371/journal.pone.0029797','http://doi.org/10.1016/j.dsr2.2010.10.029'], instance: "crossref")
  #      Lagotto.works(ids: 'http://doi.org/10.1371/journal.pone.0025110', instance: "plos")
  #      ids = ["http://doi.org/10.1371/journal.pone.0029797","http://doi.org/10.1371/journal.pone.0029798"]
  #      Lagotto.works(ids: ids, instance: "plos")
  #      Lagotto.works(ids: '10.4081/audiores.2013.e1', key: ENV['PKP_API_KEY'], instance: "pkp")
  #
  #      # Search by source
  #      Lagotto.works(source: 'twitter', instance: "crossref")
  #      Lagotto.works(instance: "crossref", per_page: 5)
  #
  #      # get by publisher
  #      Lagotto.works(publisher: 311)
  def self.works(ids: nil, type: nil, source: nil, publisher: nil,
            order: nil, per_page: 50, page: 1, instance: 'plos', key: nil,
            options: nil, verbose: false)

    test_length(source)
    type_check(page, Fixnum)
    type_check(per_page, Fixnum)
    # test_values('id_type', id_type, ['doi','pmid','pmcid','mendeley_uuid'])
    # test_values('instance', instance, ['plos','crossref','copernicus','elife','pensoft','pkp'])

    url = pick_url(instance)
    ids = join_ids(ids)
    Request.new(url, 'works', ids, type, source,
      publisher, order, per_page, page, key, options, verbose).perform
  end

  ##
  # Make a `/works` route request by source
  #
  # @param source [String] One source. To get many sources, make many calls.
  # @param per_page [String] Items per page
  # @param page [String] Page to retrieve
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param key [String] API key
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    # a single source
  #    Lagotto.works_sources(source: 'twitter', per_page: 2)
  #    Lagotto.works_sources(source: 'mendeley', per_page: 3)
  #    Lagotto.works_sources(source: 'facebook', per_page: 1)

  #    # many sources
  #    sources = ['facebook','twitter','mendeley']
  #    sources.collect { |x| Lagotto.works_sources(source: x, per_page: 1) }
  def self.works_sources(source: nil, per_page: 50, page: 1, instance: 'plos', key: nil,
    options: nil, verbose: false)

    url = pick_url(instance)
    Request.new(url, 'works', nil, nil, source,
      nil, nil, per_page, page, key, options, verbose).perform
  end

  ##
  # Make a `/events` route request
  #
  # @!macro lagotto_params
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #   # Get events by DOI
  #   ## swap out the key and instance to change provider
  #   Lagotto.events(ids: '10.1371/journal.pone.0029797', instance: "crossref")
  #   Lagotto.events(ids: ['10.1371/journal.pone.0029797','10.1016/j.dsr2.2010.10.029'], instance: "crossref")
  #   Lagotto.events(ids: '10.4081/audiores.2013.e1', instance: "pkp")
  #   Lagotto.events(ids: "10.1371/journal.pone.0029797")
  #   ids = ["10.1371/journal.pone.0029797","10.1371/journal.pone.0029798"]
  #   Lagotto.events(ids: ids)
  #
  #   # Search by source
  #   Lagotto.events(source: 'twitter', instance: "crossref")
  #   Lagotto.events(instance: "crossref", per_page: 5)
  #
  #   # Get data by publisher
  #   ids = Faraday.new(:url => "http://api.crossref.org/members").get
  #   json = MultiJson.load(ids.body)
  #   ids = json['message']['items'].collect { |p| p['id'] }
  #   Lagotto.events(publisher: ids[0])
  def self.events(ids: nil, type: nil, source: nil, publisher: nil,
    order: nil, per_page: 50, page: 1, instance: 'plos', key: nil,
    options: nil, verbose: false)

    test_length(source)
    type_check(page, Fixnum)
    type_check(per_page, Fixnum)
    url = pick_url(instance)
    ids = join_ids(ids)
    Request.new(url, 'events', ids, type, source,
      publisher, order, per_page, page, key, options, verbose).perform
  end

  ##
  # Make a `/publishers` route request
  #
  # @param id [Fixnum] Publisher id
  # @param page [String] Page to retrieve
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.publishers()
  #    Lagotto.publishers(id: 340)
  #    Lagotto.publishers(instance: 'crossref')
  def self.publishers(id: nil, page: 1, instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance) + '/publishers'
    if !id.nil?
      url = url + '/' + id.to_s
    end
    BasicRequest.new(url, '', nil, options, verbose, {}).perform
  end

  ##
  # Make a `/groups` route request
  #
  # @param id [Fixnum] Group id
  # @param page [String] Page to retrieve
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.groups()
  #    Lagotto.groups(instance: 'crossref')
  #    Lagotto.groups(id: 'recommended', instance: 'crossref')
  def self.groups(id: nil, page: 1, instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance) + '/groups'
    if !id.nil?
      url = url + '/' + id.to_s
    end
    BasicRequest.new(url, '', nil, options, verbose, {}).perform
  end

  ##
  # Make a `/references` route request
  #
  # Returns list of references for a particular work, source and/or relation_type
  #
  # @param work_id [String] Work ID
  # @param work_ids [String] Work IDs
  # @param q [String] Query for ids
  # @param relation_type_id [String] Relation_type ID
  # @param source_id [String] Source ID
  # @param page [String] Page number
  # @param per_page [String] Results per page (0-1000), Default: 1000
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.references(per_page: 5, instance: 'crossref')
  #    Lagotto.references(work_id: 'http://doi.org/10.6084/M9.FIGSHARE.1598061', instance: 'crossref')
  def self.references(work_id: nil, work_ids: nil, q: nil, relation_type_id: nil,
    source_id: nil, page: 1, per_page: 1000,
    instance: 'plos', options: nil, verbose: false)

    url = pick_url(instance)
    args = { work_id: work_id, work_ids: work_ids, q: q,
      relation_type_id: relation_type_id, source_id: source_id,
      page: page, per_page: per_page }
    BasicRequest.new(url, 'references', nil, options, verbose, args).perform
  end

  ##
  # Make a `/work_types` route request
  #
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.work_types()
  #    Lagotto.work_types(instance: 'crossref')
  def self.work_types(instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance)
    BasicRequest.new(url, 'work_types', nil, options, verbose, {}).perform
  end

  ##
  # Make a `/docs` route request
  #
  # @param id [Fixnum] Document id
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.docs()
  #    Lagotto.docs(instance: 'crossref')
  #    Lagotto.docs(id: 'counter', instance: 'crossref')
  def self.docs(id: nil, instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance) + '/docs'
    if !id.nil?
      url = url + '/' + id.to_s
    end
    BasicRequest.new(url, '', nil, options, verbose, {}).perform
  end

  ##
  # Make a `/relation_types` route request
  #
  # @param id [Fixnum] Relation type id
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.relation_types()
  #    Lagotto.relation_types(instance: 'crossref')
  #    Lagotto.relation_types(id: 'corrects', instance: 'crossref')
  def self.relation_types(id: nil, instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance) + '/relation_types'
    if !id.nil?
      url = url + '/' + id.to_s
    end
    BasicRequest.new(url, '', nil, options, verbose, {}).perform
  end

  ##
  # Make a `/sources` route request
  #
  # @param id [Fixnum] Source id
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.sources(instance: 'datacite')
  #    Lagotto.sources(instance: 'crossref')
  #    Lagotto.sources(id: 'reddit', instance: 'crossref')
  def self.sources(id: nil, instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance) + '/sources'
    if !id.nil?
      url = url + '/' + id.to_s
    end
    BasicRequest.new(url, '', nil, options, verbose, {}).perform
  end

  ##
  # Make a `/recommendations` route request
  #
  # @param id [Fixnum] A work id
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #    Lagotto.recommendations(id: '10.1371/journal.ppat.1005123')
  #    Lagotto.recommendations(id: '10.1371/journal.pgen.1005625')
  def self.recommendations(id:, instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance) + '/works/' + id.to_s + '/recommendations'
    BasicRequest.new(url, '', nil, options, verbose, {}).perform
  end

  ##
  # Make a `/status` route request
  #
  # @param key [String] API key
  # @param instance [String] One of plos, crossref, pkp, elife, copernicus, pensoft
  # @param verbose [Boolean] Print request headers to stdout. Default: false
  # @!macro lagotto_options
  # @return [Hash] A hash
  #
  # @example
  #      require 'lagotto-rb'
  #      Lagotto.status(key: ENV['PLOS_API_KEY'])
  #      Lagotto.status(key: ENV['PLOS_API_KEY'])
  def self.status(key: nil, instance: 'plos', options: nil, verbose: false)
    url = pick_url(instance)
    BasicRequest.new(url, 'status', key, options, verbose, {}).perform
  end

end
